### PR TITLE
Gen4: subquery support

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This pull request adds support for subqueries (both correlated and uncorrelated). This pull request aims for v3 parity.

## Related Issue(s)

Related to #7280 

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/abstract/operator.go [hunk 1]
```diff
@@ -29,12 +29,16 @@ type (
 	//     while still preserving the results
 	//	*  LeftJoin - A left join. These can't be evaluated in any order, so we keep them separate
 	//	*  Join - A join represents inner join.
+	//  *  SubQuery - A represents a query that encapsulates one or more sub-queries (SubQueryInner).
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/abstract/operator.go:32]: `- A represents a query ...` -> `- represents a query ...`

File: go/vt/vtgate/planbuilder/abstract/operator.go [hunk 2]
```diff
@@ -29,12 +29,16 @@ type (
...
 	Operator interface {
 		// TableID returns a TableSet of the tables contained within
 		TableID() semantics.TableSet
 
 		// PushPredicate pushes a predicate to the closest possible operator
 		PushPredicate(expr sqlparser.Expr, semTable *semantics.SemTable) error
+
+		// UnsolvedPredicates returns any predicates that have dependencies from the outside
```
Comment (systay) [inline go/vt/vtgate/planbuilder/abstract/operator.go:40]: this comment could be worked on. not easy to understand what it means

File: go/vt/vtgate/planbuilder/abstract/operator.go [hunk 3]
```diff
@@ -140,9 +161,30 @@ func CreateOperatorFromSelect(sel *sqlparser.Select, semTable *semantics.SemTabl
...
+		}
+	}
+	if resultantOp == nil {
+		return op, nil
+	}
+	resultantOp.Outer = op
+	return resultantOp, nil
+}
+
+func addRelationInformation(semTable *semantics.SemTable, expr sqlparser.Expr) {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/abstract/operator.go:174]: rename to addColumnEquality

File: go/vt/vtgate/planbuilder/abstract/subquery.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	ts := s.Outer.TableID()
+	for _, inner := range s.Inner {
+		ts = ts.Merge(inner.Inner.TableID())
+	}
+	return ts
+}
+
+// PushPredicate implements the Operator interface
+func (s *SubQuery) PushPredicate(expr sqlparser.Expr, semTable *semantics.SemTable) error {
+	panic("implement me")
```
Comment (systay) [inline go/vt/vtgate/planbuilder/abstract/subquery.go:52]: return error instead

File: go/vt/vtgate/planbuilder/rewrite_test.go [hunk 5]
```diff
@@ -182,3 +182,51 @@ func TestSemTableDependenciesAfterExpandStar(t *testing.T) {
 		})
 	}
 }
+
+func TestSubqueryRewrite(t *testing.T) {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/rewrite_test.go:186]: Test cases with `[NOT] EXISTS` should be added here

File: go/vt/vtgate/planbuilder/jointree.go [hunk 6]
```diff
@@ -77,10 +77,15 @@ type (
 		leftJoins []*outerTable
 
 		// these fields are set if a vindex will be used for this route
-		currentCost      cost // currentCost tracks the cost of the chosen access method
-		vindex           vindexes.Vindex
-		vindexValues     []sqltypes.PlanValue
+		currentCost cost // currentCost tracks the cost of the chosen access method
```
Comment (systay) [inline go/vt/vtgate/planbuilder/jointree.go:80]: we could collect all the vindex selections into a common struct, it is clear that they belong together, and to make it easy to clear out the selected opcode and vindex in `reset()`

File: go/vt/vtgate/planbuilder/route_planning_test.go [hunk 7]
```diff
@@ -55,6 +55,7 @@ func selectScatter(solved semantics.TableSet, keyspace *vindexes.Keyspace) *rout
 }
 
 func TestMergeJoins(t *testing.T) {
+	t.Skip("?")
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/route_planning_test.go:58]: do we want to keep skipping this test?

File: go/vt/vtgate/planbuilder/jointree.go [hunk 8]
```diff
@@ -92,15 +97,50 @@ type (
...
+		argName  string
+	}
 )
 
+func (s *subqueryTree) tableID() semantics.TableSet {
+	return s.inner.tableID() | s.outer.tableID()
+}
+
+func (s *subqueryTree) cost() int {
+	panic("implement me")
```
Comment (systay) [inline go/vt/vtgate/planbuilder/jointree.go:125]: need a real implementation for this. I think it is `cost(outer)+cost(inner)`, but I'm not sure

File: go/vt/vtgate/planbuilder/jointree.go [hunk 9]
```diff
@@ -92,15 +97,50 @@ type (
...
+		inner:    s.inner.clone(),
+		opcode:   s.opcode,
+		argName:  s.argName,
+	}
+	*result.subquery = *s.subquery
+	return result
+}
+
+func (s *subqueryTree) pushOutputColumns([]*sqlparser.ColName, *semantics.SemTable) ([]int, error) {
+	panic("implement me")
```
Comment (systay) [inline go/vt/vtgate/planbuilder/jointree.go:141]: return error instead

File: go/vt/vtgate/semantics/semantic_state.go [hunk 10]
```diff
@@ -495,7 +537,7 @@ func (s *scope) addTable(info TableInfo) error {
 // IsOverlapping returns true if at least one table exists in both sets
 func (ts TableSet) IsOverlapping(b TableSet) bool { return ts&b != 0 }
 
-// IsSolvedBy returns true if all of `ts` is contained in `b`
+// IsSolvedBy returns true if all of `TS` is contained in `b`
```
Comment (frouioui) [inline go/vt/vtgate/semantics/semantic_state.go:540]: `TS` -> `ts`

**Review by frouioui — CHANGES_REQUESTED** [review]

File: go/vt/vtgate/planbuilder/jointree.go [hunk 11]
```diff
@@ -597,12 +641,31 @@ func (rp *routeTree) planIsExpr(node *sqlparser.IsExpr) (bool, error) {
...
+	val, err := rp.makePlanValue(ctx, vdValue)
 	if err != nil || val == nil {
 		return false, err
 	}
 
-	return rp.haveMatchingVindex(node, column, *val, equalOrEqualUnique, justTheVindex), err
+	return rp.haveMatchingVindex(node, vdValue, column, *val, equalOrEqualUnique, justTheVindex), err
+}
+
+func (rp *routeTree) makePlanValue(ctx optimizeContext, n sqlparser.Expr) (*sqltypes.PlanValue, error) {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/jointree.go:653]: add a comment

File: go/vt/vtgate/planbuilder/rewrite.go [hunk 12]
```diff
@@ -28,16 +29,17 @@ type (
 		proceed   bool
 		tblColMap map[*sqlparser.AliasedTableExpr]sqlparser.SelectExprs
 	}
-	starRewriter struct {
-		err      error
-		semTable *semantics.SemTable
+	rewriter struct {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/rewrite.go:32]: I don't love this name. we already have other rewriters in the system - what makes this _the_ rewriter?

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 13]
```diff
@@ -147,37 +152,43 @@ func newBuildSelectPlan(sel *sqlparser.Select, reservedVars *sqlparser.ReservedV
 	return plan, nil
 }
 
-func optimizeQuery(opTree abstract.Operator, reservedVars *sqlparser.ReservedVars, semTable *semantics.SemTable, vschema ContextVSchema) (queryTree, error) {
+type optimizeContext struct {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/route_planning.go:155]: what do you think about calling this `planningContext` instead?

Comment (systay) [reply go/vt/vtgate/planbuilder/route_planning.go:155]: never mind

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 14]
```diff
@@ -186,11 +197,84 @@ func optimizeQuery(opTree abstract.Operator, reservedVars *sqlparser.ReservedVar
 			inner: treeInner,
 			alias: op.Alias,
 		}, nil
+	case *abstract.SubQuery:
```
Comment (systay) [inline go/vt/vtgate/planbuilder/route_planning.go:200]: extract method

File: go/vt/vtgate/planbuilder/abstract/operator.go (hunk 2 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/abstract/operator.go]: I changed it in ea9871c3f1, to:

```
// UnsolvedPredicates returns any predicates that have dependencies on the given Operator and
// on the outside of it (a parent Select expression, any other table not used by Operator, etc).
```

File: go/vt/vtgate/planbuilder/abstract/operator.go (hunk 3 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/abstract/operator.go]: fixed through ea9871c3f1

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 15]
```diff
@@ -708,44 +803,53 @@ func createRoutePlan(table *abstract.QueryTable, solves semantics.TableSet, rese
...
-		if !isRoute {
-			return true, nil
-		}
-		if leftDep.IsSolvedBy(rb.qtable.TableID) {
-			for _, vindex := range rb.vtable.ColumnVindexes {
-				sC, isSingle := vindex.Vindex.(vindexes.SingleColumn)
-				if isSingle && vindex.Columns[0].Equal(left.Name) {
-					singCol = sC
-					return false, io.EOF
+	for _, expr := range ctx.semTable.GetExprAndEqualities(exp) {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/route_planning.go:822]: needs comments that explain wth is going on

File: go/vt/vtgate/planbuilder/jointree.go (hunk 8 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/jointree.go]: Fixed in f83e5cbb98

File: go/vt/vtgate/planbuilder/jointree.go (hunk 9 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/jointree.go]: Fixed in f83e5cbb98

File: go/vt/vtgate/planbuilder/abstract/subquery.go (hunk 4 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/abstract/subquery.go]: Fixed in f83e5cbb98

File: go/vt/vtgate/planbuilder/testdata/filter_cases.txt [hunk 16]
```diff
@@ -1236,7 +1207,7 @@ Gen4 plan same as above
 }
 Gen4 plan same as above
 
-# subquery
+# derived table
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/filter_cases.txt:1210]: what? no, this is not a derived table, this is a subquery

File: go/vt/vtgate/planbuilder/route_planning.go (hunk 15 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/route_planning.go:826]: fixed in 710f421513

File: go/vt/vtgate/planbuilder/jointree.go (hunk 11 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/jointree.go:658]: fixed in 710f421513

File: go/vt/vtgate/planbuilder/jointree.go (hunk 6 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/jointree.go:80]: in the next PR

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/planbuilder/jointree_transformers.go [hunk 17]
```diff
@@ -96,6 +113,37 @@ func transformJoinPlan(n *joinTree, semTable *semantics.SemTable, processing *po
...
+		return false
+	}
+	sq := &sqlparser.Subquery{Select: subqSelect}
+	node = sq
+
+	// if the subquery is in an EXISTS, e.g. "__sq_has_values1"
+	// then we encapsulate the subquery in an exists expression.
+	if strings.HasPrefix(argName, string(sqlparser.HasValueSubQueryBaseName)) {
+		node = &sqlparser.ExistsExpr{Subquery: sq}
+	}
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/jointree_transformers.go:141]: ignore this.

File: go/vt/vtgate/planbuilder/rewrite.go [hunk 18]
```diff
@@ -101,13 +103,50 @@ func expandTableColumns(tables []semantics.TableInfo, starExpr *sqlparser.StarEx
...
 
-func expandStar(sel *sqlparser.Select, semTable *semantics.SemTable) (*sqlparser.Select, error) {
-	// TODO we could store in semTable whether there are any * in the query that needs expanding or not
-	sr := &starRewriter{semTable: semTable}
+func (r *rewriter) subqueryRewrite(cursor *sqlparser.Cursor) bool {
+	switch node := cursor.Node().(type) {
+	case *sqlparser.ExistsExpr:
+		semTableSQ, found := r.semTable.SubqueryRef[node.Subquery]
+		if !found {
+			// should never happen
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vtgate/planbuilder/rewrite.go:111]: ignore this.

File: go/vt/vtgate/engine/fuzz_flaky_test.go [hunk 19]
```diff
@@ -1,3 +1,6 @@
+//go:build gofuzz
+// +build gofuzz
+
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vtgate/engine/fuzz_flaky_test.go]: ignore this.

File: go/vt/vtgate/planbuilder/rewrite.go (hunk 12 above)
Comment (frouioui) [reply go/vt/vtgate/planbuilder/rewrite.go:32]: @systay, right, nothing makes it _the_ one and only rewriter. How about `postSemanticAnalysisRewriter` or something similar?

**Review by frouioui — APPROVED** [review]

**Pull request merged**